lsp/nls/src/requests/goto.rs

```diff
@@ -100,15 +105,36 @@ pub fn handle_references(
 
     // The "references" of a symbol are all the usages of its definitions,
     // so first find the definitions and then find their usages.
-    let mut def_locs = server
-        .lookup_term_by_position(pos)?
+    let term = server.lookup_term_by_position(pos)?;
+    let mut def_locs = term
         .and_then(|term| get_defs(term, ident, server))
         .unwrap_or_default();
 
     // Maybe the position is pointing straight at the definition already.
     // In that case, def_locs won't have the definition yet; so add it.
-    def_locs.extend(server.lookup_ident_by_position(pos)?);
+    if let Some(id) = server.lookup_ident_by_position(pos)? {
+        def_locs.push(id);
+        if let Some(parent) = term {
+            // If `id` is a field name in a record, we can search through cousins
+            // to find more definitions.
+            if matches!(parent.as_ref(), Term::RecRecord(..) | Term::Record(_)) {
+                let def = DefWithPath {
+                    ident: id,
+                    value: None,
+                    path: vec![id.ident],
+                    parent_record: Some(parent.clone()),
+                    metadata: None,
+                };
+                let resolver = FieldResolver::new(server);
+                let cousins = resolver.get_cousin_defs(&def);
+                def_locs.extend(cousins.into_iter().map(|(loc, _)| loc))
+            }
+        }
+    }
 
+    // TODO: This usage map is based only on static scoping, and not on our "extended"
+    // scopes that we build up dynamically based on merges. Improving this probably
+    // requires building the extended scopes at static analysis time.
```

Comment:
What exactly do you mean by that? We don't have dynamic scoping in Nickel, so you can't use a field that isn't part of the record, even if it will be brought by merging later. Or are you talking about context completion, e.g. in the case `{foo | String} & { [.] }` you would like to provide completion for `foo`?

Reply:
For context completion this already works, but one example that doesn't currently work is a "goto references" request for the first `foo` in 

```
{ foo = 1 } & { bar = foo }
```

Reply:
Sorry, that's what I get for trying to minimize an example without actually checking it :sweat_smile:. But yes, currently we go to only the definition on the RHS whereas I'd like to go to both definitions (GotoDefinition can have multiple responses). Also, when asking for References on the LHS foo, I'd like to respond with the RHS foo.

Reply:
Ah, I didn't even know goto definition could show multiple locations (and haven't noticed on Nickel). Now it makes sense, thanks!